python_modules/dagster/docs/intro_tutorial/types.rst

```diff
@@ -0,0 +1,233 @@
+    )
+    def sum_solid(num):
+        sum_df = num.copy()
+        sum_df['sum'] = sum_df['num1'] + sum_df['num2']
+        return sum_df
+
+The type metadata now appears in dagit and the system will ensure the input and output to this
+solid are indeed data frames.
+
+Input Schema
+^^^^^^^^^^^^
+
+This solid as defined is only expressed in terms of an in-memory object; it says nothing about
+how this data should be sourced from or materialized to disk. This is where the notion of
+input and output schemas come into play. Once the user provides those she is able to use
+the configuration language in order to parameterize the computation.
+
+Let us now add the input schema:
+
+.. code-block:: py
+
+    def define_path_dict_field():
+        return Field(Dict({'path': Field(Path)}))
+
+    def define_csv_dict_field():
+        return Field(
+            Dict(
+                {
+                    'path': Field(Path), 
+                    'sep': Field(
+                        String,
+                        is_optional=True,
+                        default_value=','
+                    )
+                }
+            )
+        )
+
+    @input_schema(
+        Selector(
+            {
+                'csv': define_csv_dict_field(),
+                'parquet': define_path_dict_field(),
+                'table': define_path_dict_field(),
+            }
+        )
+    )
+    def dataframe_input_schema(config_value):
+        # config_value is dictionary with a single element
+        file_type, file_options = list(config_value.items())[0]
+
+        if file_type == 'csv':
+            path = file_options['path']
+            del file_options['path']
+            return pd.read_csv(path, **file_options)
+        elif file_type == 'parquet':
+            return pd.read_parquet(file_options['path'])
+        elif file_type == 'table':
+            return pd.read_table(file_options['path'])
+        else:
+            raise DagsterInvariantViolationError(
+                'Unsupported file_type {file_type}'.format(
+                    file_type=file_type
+                )
+            )
+
+Any input schema is define by a decorated function with a single argument. The argument is the
+format the input schema takes. In this case it is a `Selector`. Selectors are used when you want
+to be able present several different options to the user but force them to select one. In this case,
+for example, it would not make much sense to allow them to say that a single input should be source
+from a csv and a parquet file: They must choose. (In other type systems this might be called an "input
+union.")
+
+Note our selector provides three keys: `csv`, `parquet`, and `table`. Each of these in turn has
+their own subfield selections. You'll note that these are `Dict` require the user to provide all
+required fields, rather than only one.
+
+The remaining step here is that the user must provide code that takes those validated configurated
```

Comment:
(small) "user" seems to be referring to 2 different people throughout this paragraph: Here, the user is the person using the dagster abstraction and they must provide the code to actually turn the config -> object (i.e. path -> dataframe). But previously "user" referred to someone who was messing with the config (for example in "Selectors are used when you want to be able present several different options to the user but force them to select one."). 

This is maybe slightly confusing?

Reply:
I'm assuming pipeline author and config author are the same human for this section